### PR TITLE
Make nestable condition factory accept conditions on super types

### PR DESCRIPTION
**Pull request by alex859**
#### Check List:
* Unit tests : YES
* Javadoc with a code example (on API only) : NA
* PR meets the [contributing guidelines](https://github.com/assertj/assertj/blob/main/CONTRIBUTING.md)

Makes it possible to create a `NestableCondition<T>` by passing in a condition on a super type of `T`.

### CODE REVIEW

File: assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java [hunk 1]
```diff
@@ -138,16 +138,16 @@ public static <ACTUAL, NESTED> Condition<ACTUAL> nestable(String descriptionPref
...
    */
   @SafeVarargs
-  public static <ACTUAL> Condition<ACTUAL> nestable(String descriptionPrefix, Condition<ACTUAL>... conditions) {
-    return new NestableCondition<>(descriptionPrefix, stream(conditions));
+  public static <ACTUAL> Condition<ACTUAL> nestable(String descriptionPrefix, Condition<? super ACTUAL>... conditions) {
+    return new NestableCondition<ACTUAL, ACTUAL>(descriptionPrefix, stream(conditions));
   }
 
-  private NestableCondition(String descriptionPrefix, Stream<Condition<NESTED>> conditions, Function<ACTUAL, NESTED> extractor) {
+  private NestableCondition(String descriptionPrefix, Stream<Condition<? super NESTED>> conditions, Function<ACTUAL, NESTED> extractor) {
```
Comment (vlsi) [inline assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java:145]: Function<a, n> should be Function<? … a, ? … n> as well

Comment (alex859) [reply assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java]: Hi @vlsi good point, thanks :+1: .

I've changed it for the function's input, but I can't find an example where changing the boundary of the function output is required :thinking: .

Comment (vlsi) [reply assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java]: `Function<ACTUAL, NESTED>` means function that `consumes ACTUAL` and `produces NESTED`.
With Java, you have to put `extends` and `super` all over the place (== in parameter declaration and return types).

For instance, `Function<A, B>` must always (except small edge cases) should be used as `Function<?  super A, ? extends B>`.
Then the method would accept `Function<.., ...>` that does not exactly match `Function<ACTUAL, NESTED>`.

See https://gist.github.com/vlsi/3e7c6a0aa3f13ec0cfc94c0f7447ceb7

File: assertj-core/src/test/java/org/assertj/core/condition/NestableCondition_matches_Test.java [hunk 2]
```diff
@@ -75,4 +79,35 @@ void should_not_match_if_any_condition_in_nested_level_does_not_match() {
...
+  void should_accept_conditions_on_supertypes() {
+    // GIVEN
+    final ValueCustomer boris = new ValueCustomer(new Name("Boris", "Johnson"),
+                                                  new Address("10, Downing Street",
+                                                              "SW1A 2AA",
+                                                              new Country("United Kingdom")),
+                                                  12);
+    Condition<ValueCustomer> valueCustomer = nestable("value customer", name(
+                                                                         first("Boris")),
+                                                  value(12));
```
Comment (vlsi) [inline assertj-core/src/test/java/org/assertj/core/condition/NestableCondition_matches_Test.java:93]: It would be nice if the lines could be wrapped. For instance, put `new ValueCustomer` on a new line, and put `new Name` on a new line as well. Currently, the code contains lots of whitespace, and it is hard to follow.

File: assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java [hunk 3]
```diff
@@ -63,6 +63,12 @@ static Condition<Customer> customer(Condition<Customer>... conditions) {
     return nestable("customer", conditions);
   }
 
+  static Condition<ValueCustomer> value(Integer expected) {
+    return verboseCondition(valueCustomer -> expected.equals(valueCustomer.value),
+      "value: " + expected,
+      valueCustomer -> " but was " + valueCustomer.value);
```
Comment (vlsi) [inline assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java:69]: This seems to duplicate the logic of `assertEquals`. Can it be de-duplicated?

File: assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java (hunk 1 above)
Comment (alex859) [reply assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java]: Thanks :+1:

File: assertj-core/src/test/java/org/assertj/core/condition/NestableCondition_matches_Test.java (hunk 2 above)
Comment (alex859) [reply assertj-core/src/test/java/org/assertj/core/condition/NestableCondition_matches_Test.java:93]: I agree, but that is the format I get applying the Eclipse format.

Comment (alex859) [reply assertj-core/src/test/java/org/assertj/core/condition/NestableCondition_matches_Test.java]: Tried to improve it.

File: assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java (hunk 3 above)
Comment (alex859) [reply assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java:69]: Not sure about this, what `assertEquals` method are you referring to?

File: assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java [hunk 4]
```diff
@@ -138,16 +138,16 @@ public static <ACTUAL, NESTED> Condition<ACTUAL> nestable(String descriptionPref
...
    */
   @SafeVarargs
-  public static <ACTUAL> Condition<ACTUAL> nestable(String descriptionPrefix, Condition<ACTUAL>... conditions) {
-    return new NestableCondition<>(descriptionPrefix, stream(conditions));
+  public static <ACTUAL> Condition<ACTUAL> nestable(String descriptionPrefix, Condition<? super ACTUAL>... conditions) {
+    return new NestableCondition<ACTUAL, ACTUAL>(descriptionPrefix, stream(conditions));
   }
 
-  private NestableCondition(String descriptionPrefix, Stream<Condition<NESTED>> conditions, Function<ACTUAL, NESTED> extractor) {
+  private NestableCondition(String descriptionPrefix, Stream<Condition<? super NESTED>> conditions, Function<? super ACTUAL, ? extends NESTED> extractor) {
```
Comment (vlsi) [inline assertj-core/src/main/java/org/assertj/core/condition/NestableCondition.java:145]: `Stream<? extends Condition<…` ?

File: assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java (hunk 3 above)
Comment (vlsi) [reply assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java:69]: `assertThat(valueCustomer.value).as("value").isEqualTo(expected)`

Comment (alex859) [reply assertj-core/src/test/java/org/assertj/core/condition/NestableConditionFixtures.java:69]: My understanding is that when creating a condition we need a predicate returning true/false, while calling `isEqual` would throw an exception when the condition is not satisfied. 

It would be nice to reuse assertions as conditions. Is there a way of doing that? (a way better than catching the exception an returning false?)

**Comment by vlsi** [comment]
`NestableCondition` seems to duplicate logic of the already existing `org.assertj.core.condition.MappedCondition`.
Have you considered reusing `MappedCondition` instead?

/cc @joel-costigliola

**Comment by vlsi** [comment]
The generified code looks good to me, however, it is sad that `nestable` duplicates logic of `MappableCondition` and it lacks support for `isEqualTo` builders.

So the users of `nestable` can't really leverage `assertj` power and they resort to `Objects.equals` and all its quircks like String and StringBuilder looking the same and returning false on equals.

**Comment by alex859** [comment]
I considered using `MappingCondition` instead of adding the `compose` method. The issue I had was that the description generated had some extra text I didn't like, e.g. in the assertion message from `NestableCondition_assertionMessage_Test#should_show_correct_error_message_with_two_nested_objects` instead of 
```
[✗] postcode: SW2A 2AA but was SW1A 2AA
```
we would get 
```
[✗] mapped
   from: <Customer> org.assertj.core.condition.Customer@68c72235
   to:   <Address> org.assertj.core.condition.Address@34f7cfd9
   then checked:
         [✗] postcode: SW2A 2AA but was SW1A 2AA
```

I agree on the lack of using the power of `assertj`, isn't that a general issue when using conditions?

**Comment by vlsi** [comment]
>I considered using MappingCondition instead of adding the compose method. The issue I had was that the description generated had some extra text I didn't like, e.g. in the assertion message from

Have you considered making MappedCondition less verbose or making the verbosity configurable?
The issue is that both `MappedCondition` and `NestedCondition` appear as implementations of `Condition`, so it makes the API hard to reason about. Personally, I often check "find all implementations" to select the one I need.

> I agree on the lack of using the power of assertj, isn't that a general issue when using conditions?

One of the ways out could be suggesting that users should resort to something like `org.assertj.core.api.HamcrestCondition.matching(org.hamcrest.Matchers.is(expected))` instead of `verboseCondition`.


An alternative way could be decouple `assertThat(...).isEqualTo()` into `AssertBuilder` and `assert verification` steps.
Of course, it might require a more elaborate integration with assertj, however, it might be in line with https://github.com/assertj/assertj/issues/2495.
In other words, it might be nice if "navigation and assertion methods" could be decoupled from "triggering the verification". Then "navigation" could be reused across regular asserts, soft asserts, and nested asserts.

**Comment by alex859** [comment]
I'm sorry I think I misunderstood your previous comment. Are you suggesting that, instead of adding `NestableCondition`, we could modify `MappedCondition` to achieve the same result and have only one of them?

Regarding `isEqual` maybe another option could be to add a condition that internally uses `StandardComparisonStrategy#areEqual`?

**Comment by vlsi** [comment]
>Are you suggesting that, instead of adding NestableCondition, we could modify MappedCondition to achieve the same result and have only one of them?

I think so.

>Regarding isEqual maybe another option could be to add a condition that internally uses StandardComparisonStrategy#areEqual?

There are multiple implementations of `StandardComparisonStrategy`.
A short-term solution might be adding something like `AreEqualCondition` instead of the current `verboseCondition`.
However, it would be sad to start adding conditions for all the methods in `ComparisonStrategy`

**Comment by alex859** [comment]
Hi @vlsi , I think you are right, the same behaviour can be achieved by having:
* a less verbose/configurable `MappedCondition`
* a configurable `AllOf` description prefix

With those changes there would be no need for `NestableCondition`, and for example `NestableConditionFixtures#address`  could be defined as something like:

```java
  static Condition<Customer> address(Condition<? super Address>... conditions) {
    return AllOf.allOf(
      "address", 
      Arrays.stream(conditions)
        .map(condition -> lessVerboseMappedCondition((Customer customer) -> customer.address, condition))
        .collect(Collectors.toList())
    );
  }
```

Thoughts?

**Comment by vlsi** [comment]
I believe "address" string should be a part of lessVerboseCondition, not an argument to allof

**Comment by alex859** [comment]
Good point. I think that some improvements can be discussed and addressed with a separate PR, the goal of this one was just to improve the current state of `NestableCondition`.

**Review by vlsi — COMMENTED** [review]
The improved generics look good to me

**Comment by joel-costigliola** [comment]
Will look at it this weekend, thanks for your patience

**Comment by alex859** [comment]
Thank you! It would be great if you could also review #2755.

**Comment by alex859** [comment]
Thank you! It would be great if you could also review #2755.

**Comment by joel-costigliola** [comment]
Integrated thanks @alex859!

**Comment by alex859** [comment]
Thank you! ☺️
